Error while building: _FORTIFY_SOURCE requires compiling with optimization

I'm trying to use melior, and I'm getting an error while building tblgen:

```
warning: tblgen@0.6.0: In file included from /nix/store/qdknxw57cwy1jkrhq7fzmiis73j42jv6-gcc-14.3.0/include/c++/14.3.0/x86_64-unknown-linux-gnu/bits/os_defines.h:39,
warning: tblgen@0.6.0:                  from /nix/store/qdknxw57cwy1jkrhq7fzmiis73j42jv6-gcc-14.3.0/include/c++/14.3.0/x86_64-unknown-linux-gnu/bits/c++config.h:683,
warning: tblgen@0.6.0:                  from /nix/store/qdknxw57cwy1jkrhq7fzmiis73j42jv6-gcc-14.3.0/include/c++/14.3.0/bits/memoryfwd.h:48,
warning: tblgen@0.6.0:                  from /nix/store/qdknxw57cwy1jkrhq7fzmiis73j42jv6-gcc-14.3.0/include/c++/14.3.0/memory:63,
warning: tblgen@0.6.0:                  from cc/lib/TableGen.hpp:14,
warning: tblgen@0.6.0:                  from cc/lib/TableGen.cpp:11:
warning: tblgen@0.6.0: /nix/store/9ly5qx1lbj0q1rcsf1dc0m9n3db6ii60-glibc-2.40-66-dev/include/features.h:422:4: error: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Werror=cpp]
warning: tblgen@0.6.0:   422 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
warning: tblgen@0.6.0:       |    ^~~~~~~
warning: tblgen@0.6.0: cc1plus: all warnings being treated as errors
error: failed to run custom build command for `tblgen v0.6.0`
```

I'm using a version of llvm that I built from source with options:

```
cmake -G Ninja ../llvm    -DLLVM_ENABLE_PROJECTS='mlir;clang'    -DLLVM_BUILD_EXAMPLES=ON    -DLLVM_TARGETS_TO_BUILD="Native"    -DCMAKE_BUILD_TYPE=Release    -DLLVM_ENABLE_ASSERTIONS=ON    -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++    -DLLVM_CCACHE_BUILD=ON -DLLVM_SOURCE_PREFIX=. -DCMAKE_INSTALL_PREFIX:PATH=/home/o/.local/software/llvm/versions/20.1.
```

I'm not sure if this is a bug with `build.rs` or if it's something I'm doing wrong.

--- a/build.rs
+++ b/build.rs
@@ -119,9 +119,32 @@ fn run() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
+fn filter_fortify_source(flags: &str) -> String {
+    flags
+        .split_whitespace()
+        .filter(|flag| !flag.starts_with("-D_FORTIFY_SOURCE"))
+        .collect::<Vec<_>>()
+        .join(" ")
+}
+
 fn build_c_library() -> Result<(), Box<dyn Error>> {
-    unsafe { env::set_var("CXXFLAGS", llvm_config(false, "--cxxflags")?) };
-    unsafe { env::set_var("CFLAGS", llvm_config(false, "--cflags")?) };
+    let raw_cxxflags = llvm_config(false, "--cxxflags")?;
+    let raw_cflags = llvm_config(false, "--cflags")?;
+
+    // The cc crate does not add -O when OPT_LEVEL=0, so -D_FORTIFY_SOURCE (which
+    // glibc requires to be paired with optimization) causes a #warning that
+    // -Werror turns into a hard error. Strip it only when compiling without
+    // optimization; otherwise keep it for the runtime hardening it provides.
+    let (cxxflags, cflags) = if env::var("OPT_LEVEL").as_deref() == Ok("0") {
+        (
+            filter_fortify_source(&raw_cxxflags),
+            filter_fortify_source(&raw_cflags),
+        )
+    } else {
+        (raw_cxxflags, raw_cflags)
+    };
+    unsafe { env::set_var("CXXFLAGS", cxxflags) };
+    unsafe { env::set_var("CFLAGS", cflags) };
 
     cc::Build::new()
         .cpp(true)

